### PR TITLE
8295303: cleanup debug agent's confusing use of EI_GC_FINISH

### PR DESCRIPTION
**Pull request by plummercj**
This PR renamed EI_GC_FINISH to EI_CLASS_UNLOAD and does some other minor cleanups related to EI_GC_FINISH.

The debug agent deals with 3 types of events: JVMTI (per the spec), JDWP (per the spec), and an event indexing that the debug agent refers to as EI (Event Index) events. We have the following EI_GC_FINISH event. The indexing is into the array of jdwp event handlers which are created when the debugger sends an EventRequest command.

Note there is no EI_CLASS_UNLOAD event that maps to the JDWP CLASS_UNLOAD event, but instead we have:

`        EI_GC_FINISH = 8,`

And these are the mappings between EI_GC_FINISH and JDWP and JVMTI events

```
  case JVMTI_EVENT_GARBAGE_COLLECTION_FINISH:
      return EI_GC_FINISH;

  case JDWP_EVENT(CLASS_UNLOAD):
      return EI_GC_FINISH;

  index2jvmti[EI_GC_FINISH - EI_min] = JVMTI_EVENT_GARBAGE_COLLECTION_FINISH;

  index2jdwp[EI_GC_FINISH - EI_min] = JDWP_EVENT(CLASS_UNLOAD);
```

So there is this odd relationship between EI_GC_FINISH and the JDWP CLASS_UNLOAD event. Note that JVMTI does not have a CLASS_UNLOAD (except for unused support in the extension mechanism), and JDWP has no GC_FINISH event.

This relationship between EI_GC_FINISH and the JDWP CLASS_UNLOAD events stems from the fact that at one point JVMTI_EVENT_GARBAGE_COLLECTION_FINISH was used to trigger the synthesizing all of JDWP CLASS_UNLOAD events for classes that unloaded during the last GC. That is no longer the case, and instead each time a JVMTI OBJECT_FREE event is triggered for a Class instance, the JDWP CLASS_UNLOAD is generated.

Since JDWP CLASS_UNLOAD maps to EI_GC_FINISH, we have the following:

`    node = getHandlerChain(EI_GC_FINISH)->first;`

By looking at this line of code, you would never guess that this is how you fetch the event handler chain for JDWP CLASS_UNLOAD EventRequests, but it is.

To clean this up I renamed EI_GC_FINISH to EI_CLASS_UNLOAD. However, that still leaves the EI_GC_FINISH to JVMTI_EVENT_GARBAGE_COLLECTION_FINISH mapping to deal with. It's not needed anymore. When we get a JVMTI_EVENT_GARBAGE_COLLECTION_FINISH, this is all we ever do with it:

```
static void JNICALL
cbGarbageCollectionFinish(jvmtiEnv *jvmti_env)
{
    LOG_CB(("cbGarbageCollectionFinish"));
    ++garbageCollected;
    LOG_MISC(("END cbGarbageCollectionFinish"));
}
```

So the event is not passed around at all, and doesn't trigger other events or mapping to a JDWP event. It is never used as an "event index". This means jvmti2EventIndex should assert if it is ever passed in, since following mapping should never be needed:

```
        case JVMTI_EVENT_GARBAGE_COLLECTION_FINISH:
            return EI_GC_FINISH; 
```

This is accomplished by simply deleting the above code, and failing through to the default error handling code.

Also no longer needed is the following entry:

    index2jvmti[EI_GC_FINISH          -EI_min] = JVMTI_EVENT_GARBAGE_COLLECTION_FINISH

For this we just assign the entry to 0, which will result in an error if the entry is ever referenced.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8295303](https://bugs.openjdk.org/browse/JDK-8295303): cleanup debug agent's confusing use of EI_GC_FINISH

### Reviewers
 * [Alex Menkov](https://openjdk.org/census#amenkov) (@alexmenkov - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10887/head:pull/10887` \
`$ git checkout pull/10887`

Update a local copy of the PR: \
`$ git checkout pull/10887` \
`$ git pull https://git.openjdk.org/jdk pull/10887/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10887`

View PR using the GUI difftool: \
`$ git pr show -t 10887`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10887.diff">https://git.openjdk.org/jdk/pull/10887.diff</a>

</details>

### CODE REVIEW

File: src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c [hunk 1]
```diff
@@ -1504,11 +1504,17 @@ eventHandler_initialize(jbyte sessionID)
...
         EXIT_ERROR(error,"Can't enable thread end events");
     }
-    error = threadControl_setEventMode(JVMTI_ENABLE,
-                                       EI_GC_FINISH, NULL);
+
+    /*
+     * GARBAGE_COLLECTION_FINISH is special since it is not tied to any handlers or an EI,
+     * so it cannot be setup using threadControl_setEventMode(). Use JVMTI API directly.
+     */
+    error = JVMTI_FUNC_PTR(gdata->jvmti,SetEventNotificationMode)
```
Comment (alexmenkov) [inline src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c:1512]: Please add space after the comma:
error = JVMTI_FUNC_PTR(gdata->jvmti, SetEventNotificationMode)

File: src/jdk.jdwp.agent/share/native/libjdwp/util.c [hunk 2]
```diff
@@ -1977,21 +1977,29 @@ eventIndexInit(void)
 }
 
 jdwpEvent
-eventIndex2jdwp(EventIndex i)
+eventIndex2jdwp(EventIndex ei)
 {
-    if ( i < EI_min || i > EI_max ) {
-        EXIT_ERROR(AGENT_ERROR_INVALID_INDEX,"bad EventIndex");
+    jdwpEvent event = 0;
+    if (ei >= EI_min || ei >= EI_max) {
```
Comment (alexmenkov) [inline src/jdk.jdwp.agent/share/native/libjdwp/util.c:1983]: Should be "(ei >= EI_min && ei <= EI_max"

File: src/jdk.jdwp.agent/share/native/libjdwp/util.c [hunk 3]
```diff
@@ -1977,21 +1977,29 @@ eventIndexInit(void)
...
 }
 
 jvmtiEvent
-eventIndex2jvmti(EventIndex i)
+eventIndex2jvmti(EventIndex ei)
 {
-    if ( i < EI_min || i > EI_max ) {
-        EXIT_ERROR(AGENT_ERROR_INVALID_INDEX,"bad EventIndex");
+    jvmtiEvent event = 0;
+    if (ei >= EI_min || ei >= EI_max) {
```
Comment (alexmenkov) [inline src/jdk.jdwp.agent/share/native/libjdwp/util.c:1996]: Should be "(ei >= EI_min && ei <= EI_max"

**Review by alexmenkov — CHANGES_REQUESTED** [review]

File: src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c (hunk 1 above)
Comment (plummercj) [reply src/jdk.jdwp.agent/share/native/libjdwp/eventHandler.c:1512]: Omitting the space is consistent with other uses of JVMTI_FUNC_PTR.

File: src/jdk.jdwp.agent/share/native/libjdwp/util.c (hunk 2 above)
Comment (plummercj) [reply src/jdk.jdwp.agent/share/native/libjdwp/util.c]: Fixed.

File: src/jdk.jdwp.agent/share/native/libjdwp/util.c (hunk 3 above)
Comment (plummercj) [reply src/jdk.jdwp.agent/share/native/libjdwp/util.c]: Fixed.

File: src/jdk.jdwp.agent/share/native/libjdwp/error_messages.c [hunk 4]
```diff
@@ -232,6 +232,7 @@ eventText(int i)
         CASE_RETURN_TEXT(EI_THREAD_START)
         CASE_RETURN_TEXT(EI_THREAD_END)
         CASE_RETURN_TEXT(EI_CLASS_PREPARE)
+        CASE_RETURN_TEXT(EI_CLASS_UNLOAD)
         CASE_RETURN_TEXT(EI_CLASS_LOAD)
```
Comment (sspitsyn) [inline src/jdk.jdwp.agent/share/native/libjdwp/error_messages.c:236]: Nit: Would it be a little better to place it after EI_CLASS_UNLOAD?

**Review by sspitsyn — APPROVED** [review]
Looks good.
Thanks,
Serguei

File: src/jdk.jdwp.agent/share/native/libjdwp/error_messages.c (hunk 4 above)
Comment (plummercj) [reply src/jdk.jdwp.agent/share/native/libjdwp/error_messages.c:236]: If you mean put EI_CLASS_UNLOAD after EI_CLASS_LOAD, I tried to keep the current ordering. If you look in util.h (bottom of this page), it used to have:
```
        EI_CLASS_PREPARE        =  7,
        EI_GC_FINISH            =  8,
        EI_CLASS_LOAD           =  9,
```
and now it has:
``` 
        EI_CLASS_PREPARE        =  7,
        EI_CLASS_UNLOAD         =  8,
        EI_CLASS_LOAD           =  9,
```
There are a couple of other places where this ordering is preserved and I tried to be consistent.

**Review by alexmenkov — APPROVED** [review]

**Comment by plummercj** [comment]
Thanks Alex and Serguei!

/integrate

File: src/jdk.jdwp.agent/share/native/libjdwp/error_messages.c (hunk 4 above)
Comment (sspitsyn) [reply src/jdk.jdwp.agent/share/native/libjdwp/error_messages.c:236]: Okay, thanks.
It makes sense then.

**Pull request closed without merging**